async-nats/src/lib.rs

```diff
@@ -501,9 +509,33 @@ impl Connector {
     pub(crate) async fn try_connect(&mut self) -> Result<(ServerInfo, Connection), io::Error> {
         let mut error = None;
 
-        for server_addr in &self.server_addrs {
-            match self.try_connect_to(server_addr).await {
-                Ok(inner) => return Ok(inner),
+        let server_addrs: Vec<ServerAddr> = self.attempts.keys().cloned().collect();
+        for server_addr in server_addrs {
+            let server_attempts = self.attempts.get_mut(&server_addr).unwrap();
+            let duration = if *server_attempts == 0 {
+                Duration::from_millis(0)
+            } else {
+                let exp: u32 = (*server_attempts - 1).try_into().unwrap_or(std::u32::MAX);
+                let max = Duration::from_secs(4);
+
+                cmp::min(Duration::from_millis(2_u64.saturating_pow(exp)), max)
+            };
+
+            *server_attempts += 1;
+            sleep(duration).await;
+
+            match self.try_connect_to(&server_addr).await {
+                Ok((info, connection)) => {
+                    for url in &info.connect_urls {
+                        self.attempts.insert(url.parse()?, 0);
+                    }
+
+                    let server_attempts = self.attempts.get_mut(&server_addr).unwrap();
+                    *server_attempts = 0;
```

Comment:
isn't line 524 `server_attempts` still alive here?

Reply:
It can't be, would be double borrow for self.